client-src/elements/utils_test.ts

```diff
@@ -346,4 +364,307 @@ go/this-is-a-test
+        feature,
+        shippingInfoUpcoming,
+        CURRENT_DATE_MS,
+        true
+      );
+      assert.equal(getBannerId(result), 'outdated-icon');
+    });
+
+    it('should return upcoming banner if upcoming, outdated, and user cannot edit', () => {
+      const feature = {...baseFeature, accurate_as_of: OUTDATED_DATE};
+      const result = getFeatureOutdatedBanner(
+        feature,
+        shippingInfoUpcoming,
+        CURRENT_DATE_MS,
+        false
+      );
+      assert.equal(getBannerId(result), 'outdated-icon');
+    });
+
+    // --- Shipped Features ---
+
+    it('should return null if feature is shipped but verified', () => {
+      const feature = {...baseFeature, accurate_as_of: VERIFIED_DATE};
+      const result = getFeatureOutdatedBanner(
+        feature,
+        shippingInfoShipped,
+        CURRENT_DATE_MS,
+        true
+      );
+      assert.isNull(result);
+    });
+
+    it('should return author banner if shipped, outdated (<9 weeks), and user can edit', () => {
+      const feature = {...baseFeature, accurate_as_of: OUTDATED_DATE};
+      const result = getFeatureOutdatedBanner(
+        feature,
+        shippingInfoShipped,
+        CURRENT_DATE_MS,
+        true
+      );
+      assert.equal(getBannerId(result), 'shipped-outdated-author');
+    });
+
+    it('should return null if shipped, outdated (<9 weeks), and user cannot edit', () => {
+      const feature = {...baseFeature, accurate_as_of: OUTDATED_DATE};
+      const result = getFeatureOutdatedBanner(
+        feature,
+        shippingInfoShipped,
+        CURRENT_DATE_MS,
+        false
+      );
+      assert.isNull(result);
+    });
+
+    it('should return "all" banner if shipped, outdated (>9 weeks), and user can edit', () => {
+      const feature = {...baseFeature, accurate_as_of: VERY_OUTDATED_DATE};
+      const result = getFeatureOutdatedBanner(
+        feature,
+        shippingInfoShipped,
+        CURRENT_DATE_MS,
+        true
+      );
+      // Author banner still takes precedence
+      assert.equal(getBannerId(result), 'shipped-outdated-author');
+    });
+
+    it('should return "all" banner if shipped, outdated (>9 weeks), and user cannot edit', () => {
+      const feature = {...baseFeature, accurate_as_of: VERY_OUTDATED_DATE};
+      const result = getFeatureOutdatedBanner(
+        feature,
+        shippingInfoShipped,
+        CURRENT_DATE_MS,
+        false
+      );
+      assert.equal(getBannerId(result), 'shipped-outdated-all');
+    });
+
+    // --- Edge Cases ---
+
+    it('should return null if not upcoming and not shipped', () => {
+      const feature = {...baseFeature, accurate_as_of: OUTDATED_DATE};
+      const result = getFeatureOutdatedBanner(
+        feature,
+        shippingInfoNeither,
+        CURRENT_DATE_MS,
+        true
+      );
+      assert.isNull(result);
+    });
+
+    it('should return a banner if accurate_as_of is missing for upcoming', () => {
```

Comment:
Suggestion: This test case name, `'should return a banner if accurate_as_of is missing for upcoming'`, is a little ambiguous. It correctly tests that a banner *is* shown, but it's not immediately obvious *why*. What do you think about renaming it to something like `'should return upcoming banner if accurate_as_of is missing'` and adding a brief comment inside the test explaining that a missing `accurate_as_of` date is considered outdated for an upcoming feature? It would make the intent clearer for future maintainers.